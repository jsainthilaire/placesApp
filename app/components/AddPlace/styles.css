.container {
    color: #8D6E63;
    display: flex;
    justify-content: center;
}

.addPlace {
    align-items: flex-end;
    background-color: #EFEBE9;
    border-radius: 38px;
    box-shadow: 0px 2px 5px rgba(0, 0, 0, 0.4);
    color: #8D6E63;
    display: flex;
    flex-direction: column;
    padding: 15px;
}

.title {
    align-self: center;
}

.saveButton {
    background-color: #8D6E63;
    box-shadow: 0 1px 1px rgba(0, 0, 0, .6);
    color: #ffffff;
    cursor: pointer;
    font-family: inherit;
    font-size: 15px;
    height: 30px;
    outline: none;
    width: 100px;
}

.saveButton:hover {
    background-color: #7d655d;
}

.inputStyles {
    border-color: #D7CCC8;
    outline-color: #D7CCC8;
    margin: 5px 0;
    margin-left: 7px;
}

.nameBox {
    composes: inputStyles;
    border-width: thin;
    height: 20px;
    padding: 4px;
    width: 200px;
}

.descriptionBox {
    composes: inputStyles;
    border-radius: 6px;
}

.label {
    align-items: center;
    display: flex;
    font-size: 20px;
}

.imageBox {
    padding: 1px;
    border-radius: 8px;
    width: 200px;
}

.imageContainer {
    display: flex;
    max-height: 100px;
}
