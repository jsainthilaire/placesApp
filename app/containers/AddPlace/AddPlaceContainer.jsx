import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { AddPlace } from 'components'
import { updateNewPlace, saveNewPlace } from 'redux/modules/places'

class AddPlaceContainer extends Component {
  constructor(props) {
    super(props)
    this.state = {
      image: null,
      imagePreviewURL: '',
    }
  }
  submitPlace = () => {
    const {
      description,
      name,
    } = this.props

    const { image } = this.state

    this.props.saveNewPlace({ description, name, image })
  }

  updateNewPlace = fieldName => (e) => {
    this.props.updateNewPlace(fieldName, e.target.value)
  }

  handleImageChange = (e) => {
    const image = e.target.files[0]
    const reader = new FileReader()

    reader.onloadend = () => {
      this.setState({
        image,
        imagePreviewURL: reader.result,
      })
    }

    reader.readAsDataURL(image)
  }

  render() {
    return (
      <AddPlace
        updateName={this.updateNewPlace('name')}
        updateDescription={this.updateNewPlace('description')}
        onImageChange={this.handleImageChange}
        imagePreviewURL={this.state.imagePreviewURL}
        submitPlace={this.submitPlace}
        description={this.props.description}
        name={this.props.name}
      />
    )
  }
}

AddPlaceContainer.propTypes = {
  description: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  updateNewPlace: PropTypes.func.isRequired,
  saveNewPlace: PropTypes.func.isRequired,
}

const mapStateToProps = ({ places }) => ({
  description: places.newPlace.description,
  name: places.newPlace.name,
})

const mapDispatchToProps = dispatch => ({
  updateNewPlace(fieldName, fieldValue) {
    dispatch(updateNewPlace(fieldName, fieldValue))
  },
  saveNewPlace(place) {
    dispatch(saveNewPlace(place))
  },
})

export default connect(mapStateToProps, mapDispatchToProps)(AddPlaceContainer)
