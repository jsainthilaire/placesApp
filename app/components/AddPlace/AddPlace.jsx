import React from 'react'
import PropTypes from 'prop-types'
import { addPlace, saveButton, descriptionBox, nameBox, label, title, imageContainer, imageBox } from './styles.css'

const AddPlace = ({
  updateName,
  updateDescription,
  submitPlace,
  onImageChange,
  description,
  name,
  imagePreviewURL,
}) => (
  <div className={addPlace}>
    <h2 className={title}>Save new Place</h2>
    <label className={label}>
      Name
      <input
        value={name}
        onChange={updateName}
        type='text'
        className={nameBox}
      />
    </label>
    <label className={label}>
      Description
      <textarea
        value={description}
        onChange={updateDescription}
        rows='4'
        cols='50'
        className={descriptionBox}
      />
    </label>
    <div className={imageContainer}>
      <input
        type='file'
        onChange={onImageChange}
      />
      <img className={imageBox} alt={name} src={imagePreviewURL} />
    </div>
    <button className={saveButton} onClick={submitPlace}>Save Place</button>
  </div>
)

AddPlace.propTypes = {
  description: PropTypes.string.isRequired,
  imagePreviewURL: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  updateName: PropTypes.func.isRequired,
  updateDescription: PropTypes.func.isRequired,
  submitPlace: PropTypes.func.isRequired,
  onImageChange: PropTypes.func.isRequired,
}

export default AddPlace
