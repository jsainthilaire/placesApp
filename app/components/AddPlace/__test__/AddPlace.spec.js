import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import noop from 'lodash'
import AddPlace from '../AddPlace'

const minProps = {
  updateName: noop,
  updateDescription: noop,
  submitPlace: noop,
  onImageChange: noop,
  description: '',
  name: '',
  imagePreviewURL: '',
}

test('AddPlace render correctly', () => {
  const wrapper = shallow(<AddPlace {...minProps} />)
  expect(toJson(wrapper)).toMatchSnapshot()
})

test('AddPlace should call submitPlace after clicking on save button', () => {
  const submitPlace = jest.fn()
  const wrapper = shallow(<AddPlace {...minProps} submitPlace={submitPlace} />)
  wrapper.find('.saveButton').simulate('click')
  expect(submitPlace).toBeCalled()
})

it('AddPlace should call updateName after changing the name input', () => {
  const updateName = jest.fn()
  const wrapper = shallow(<AddPlace {...minProps} updateName={updateName} />)
  wrapper.find('.nameBox').simulate('change', { target: { value: 'name changed!' } })

  expect(updateName).toBeCalled()
  expect(toJson(wrapper)).toMatchSnapshot()
})

it('AddPlace should call updateDescription after changing the description textarea', () => {
  const updateDescription = jest.fn()
  const wrapper = shallow(<AddPlace {...minProps} updateDescription={updateDescription} />)
  wrapper.find('.descriptionBox').simulate('change', { target: { value: 'description is here' } })

  expect(updateDescription).toBeCalled()
})
