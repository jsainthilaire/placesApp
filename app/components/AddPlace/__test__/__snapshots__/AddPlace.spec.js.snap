// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddPlace render correctly 1`] = `
<div
  className="addPlace"
>
  <h2
    className="title"
  >
    Save new Place
  </h2>
  <label
    className="label"
  >
    Name
    <input
      className="nameBox"
      onChange={[Function]}
      type="text"
      value=""
    />
  </label>
  <label
    className="label"
  >
    Description
    <textarea
      className="descriptionBox"
      cols="50"
      onChange={[Function]}
      rows="4"
      value=""
    />
  </label>
  <div
    className="imageContainer"
  >
    <input
      onChange={[Function]}
      type="file"
    />
    <img
      alt=""
      className="imageBox"
      src=""
    />
  </div>
  <button
    className="saveButton"
    onClick={[Function]}
  >
    Save Place
  </button>
</div>
`;

exports[`AddPlace should call updateName after changing the name input 1`] = `
<div
  className="addPlace"
>
  <h2
    className="title"
  >
    Save new Place
  </h2>
  <label
    className="label"
  >
    Name
    <input
      className="nameBox"
      onChange={[Function]}
      type="text"
      value=""
    />
  </label>
  <label
    className="label"
  >
    Description
    <textarea
      className="descriptionBox"
      cols="50"
      onChange={[Function]}
      rows="4"
      value=""
    />
  </label>
  <div
    className="imageContainer"
  >
    <input
      onChange={[Function]}
      type="file"
    />
    <img
      alt=""
      className="imageBox"
      src=""
    />
  </div>
  <button
    className="saveButton"
    onClick={[Function]}
  >
    Save Place
  </button>
</div>
`;
